[DNM][WIP] catalog/descs: refactor a bit

ajwerner opened the pull request.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/65901)
<!-- Reviewable:end -->

ajwerner committed:
sql: remove a catalogkv usage

Release note: None

ajwerner committed:
sql: refactor away a couple of small descriptor resolution functions

Release note: None

ajwerner committed:
sql: clean up some resolver and flags logic

This makes the code a tad clearer. The fallback is no longer needed because the
virtual descriptors are underneath the Descriptors(). There is an inconsistency
here when it comes to schema usage. For the types we still use a resolver
method, seemingly because it does something good resolving annotating the type
name and checking its parent schema for USAGE. We should fix this inconsistency
at some point.

Release note: None

ajwerner committed:
sql/catalog/resolver: remove useless code

I'm pretty sure that this code does nothing.

Release note: None

ajwerner committed:
sql: set required flag to true

The "ByID" lookups imply required at a lower level currently, but it's
confusing to set it to false here. All callers treat it as true.

Release note: None

ajwerner committed:
sql/catalog/resolver,tree: move ResolveTargetObject from tree to resolver

Had to rename a variable along the way.

Release note: None

ajwerner committed:
sql/catalog: break catalogkeys dep on systemschema via catconstants

Release note: None

ajwerner committed:
sql/catalog/resolver,tree: move ResolveTarget to catalog/resolver from tree

Release note: None

ajwerner committed:
sql/catalog/resolver,tree: move most of name resolution out of tree

Release note: None
